cpp/include/cuspatial/experimental/detail/derive_trajectories.cuh

```diff
@@ -0,0 +1,149 @@
+#include <thrust/reduce.h>
+#include <thrust/sort.h>
+
+#include <cstdint>
+
+namespace cuspatial {
+
+namespace detail {
+
+template <typename IdInputIt,
+          typename PointInputIt,
+          typename TimestampInputIt,
+          typename IdOutputIt,
+          typename PointOutputIt,
+          typename TimestampOutputIt>
+void order_trajectories(IdInputIt ids_first,
+                        IdInputIt ids_last,
+                        PointInputIt points_first,
+                        TimestampInputIt timestamps_first,
+                        IdOutputIt ids_out_first,
+                        PointOutputIt points_out_first,
+                        TimestampOutputIt timestamps_out_first,
+                        rmm::cuda_stream_view stream,
+                        rmm::mr::device_memory_resource* mr)
+{
+  using id_type        = typename std::iterator_traits<IdInputIt>::value_type;
+  using timestamp_type = typename std::iterator_traits<TimestampInputIt>::value_type;
+  using tuple_type     = thrust::tuple<id_type, timestamp_type>;
+
+  auto key_comparator = [] __device__(tuple_type const& lhs, tuple_type const& rhs) {
+    auto lhs_id = thrust::get<0>(lhs);
+    auto rhs_id = thrust::get<0>(rhs);
+    auto lhs_ts = thrust::get<1>(lhs);
+    auto rhs_ts = thrust::get<1>(rhs);
+    return (lhs_id < rhs_id) || ((lhs_id == rhs_id) && (lhs_ts < rhs_ts));
+  };
+
+  auto keys_first     = thrust::make_zip_iterator(ids_first, timestamps_first);
+  auto keys_out_first = thrust::make_zip_iterator(ids_out_first, timestamps_out_first);
+
+  std::size_t temp_storage_bytes = 0;
+  cub::DeviceMergeSort::SortPairsCopy(nullptr,
+                                      temp_storage_bytes,
+                                      keys_first,
+                                      points_first,
+                                      keys_out_first,
+                                      points_out_first,
+                                      std::distance(ids_first, ids_last),
+                                      key_comparator);
+
+  // allocate temp storage
+  auto temp_storage = rmm::device_buffer(temp_storage_bytes, stream, mr);
+
+  cub::DeviceMergeSort::SortPairsCopy(temp_storage.data(),
+                                      temp_storage_bytes,
+                                      keys_first,
+                                      points_first,
+                                      keys_out_first,
+                                      points_out_first,
+                                      std::distance(ids_first, ids_last),
+                                      key_comparator);
+}
+
+}  // namespace detail
+
+template <typename IdInputIt,
+          typename PointInputIt,
+          typename TimestampInputIt,
+          typename IdOutputIt,
+          typename PointOutputIt,
+          typename TimestampOutputIt,
+          typename OffsetType>
+std::unique_ptr<rmm::device_vector<OffsetType>> derive_trajectories(
```

Comment:
Let's standardize on returning `device_uvector` when we need to allocate output vectors.

Reply:
Done.